data/modules/Assassination/Assassination.lua

```diff
@@ -284,35 +285,19 @@ local onEnterSystem = function (ship)
 				if mission.due > Game.time then
 					if mission.location:IsSameSystem(syspath) then -- spawn our target ship
 						local station = Space.GetBody(mission.location.bodyIndex)
-						local shiptype = ShipDef[mission.shipid]
-						local default_drive = shiptype.hyperdriveClass
-						local laserdefs = utils.build_array(pairs(Equipment.laser))
-						table.sort(laserdefs, function (l1, l2) return l1.price < l2.price end)
-						local laserdef = laserdefs[mission.danger]
-						local count = default_drive ^ 2
-
-						mission.ship = Space.SpawnShipDocked(mission.shipid, station)
-						if mission.ship == nil then
-							return -- TODO
-						end
 
-						mission.ship:SetLabel(mission.shipregid)
+						local plan = ShipBuilder.MakePlan(AssassinationTargetShip, HullConfig.GetHullConfigs()[mission.shipid], mission.threat)
+						assert(plan)
 
-						mission.ship:AddEquip(Equipment.misc.atmospheric_shielding)
-						local engine = Equipment.hyperspace['hyperdrive_'..tostring(default_drive)]
-						mission.ship:AddEquip(engine)
-						mission.ship:AddEquip(laserdef)
-						mission.ship:AddEquip(Equipment.misc.shield_generator, mission.danger)
+						local target = Space.SpawnShipDocked(mission.shipid, station)
 
-						mission.ship:GetComponent('CargoManager'):AddCommodity(Commodities.hydrogen, count)
-
-						if mission.danger > 2 then
-							mission.ship:AddEquip(Equipment.misc.shield_energy_booster)
+						if not target then
+							return -- TODO
 						end
 
-						if mission.danger > 3 then
-							mission.ship:AddEquip(Equipment.misc.laser_cooling_booster)
-						end
+						ShipBuilder.ApplyPlan(target, plan)
```

Comment:
Discussion;

To me, this seems a bit of a weird flow - you create a Hull, then generate a Plan for that Hull, then create an actual Ship (with, presumably, some set of default equipment), and finally, overwrite the Ship with the Plan?

I guess you need to know the Hull for the advert, but why bother with the Plan? Or there a plan (!) to apply a ship Plan to multiple Ships in future? (currently that's not the case)

I'd also advocate for calling  `ShipBuilder.MakeShipDocked()`, and extend the `template` to include an optional Hull instead of essentially duplicating the code.

Reply:
Hull (abstract concept/specification of a ship) -> Plan (abstract list of equipment for the ship) -> Ship (concrete reified ship object that has been spawned in the game world) -> ApplyPlan (add equipment to the currently-empty ship).

Note that specifically the code doesn't "create" the hull configuration, that's the result of a content author. We select a valid hull configuration from a pool of human-authored configurations, then create a plan to populate it with equipment, then make a ship in the world and apply the plan to it to give the concrete ship object the equipment contained in the Plan object.

I specifically have a separate Plan object from a reified instance of a Ship specifically because there are corner cases where you need to be able to "outfit" a ship that doesn't actually exist currently. For example, the main menu.

If I remember correctly, this callsite is a bit awkward because we need to know what type of ship we're going to spawn "up front". However, the "public API" of ShipBuilder is actually SelectHull->MakePlan->ApplyPlan - all of the MakeShipXXX methods are just convenience wrappers which make a concrete ship object internally.

I could extend MakeShipDocked to support passing in a hull config (it actually does effectively currently support this) but my thinking for this module was that I'd rather have the entire ship Plan be generated "ahead of time" at advert creation so that the mission description can have greater depth (and the rewards be more accurate).

Reply:
Sorry, yes I realised the `Hull` is merely selected, not created (bad wording on my behalf).

But I didn't realise that a spawned ship will always be empty, so that's one of my concerns gone. 

Still not dreadfully happy about code duplication but I understand wanting to expose more information to the mission in this case.  Nor the required coupling between the spawned-ship and the Plan relying on the same Hull (I mean, of course it needs it, but I'm not happy about the user, ie, developer, having to drive that) 

I guess a possible change could be that SpawnShip takes not only the Hull but also the Plan. Or perhaps it would only need the Plan? (I am currently under the assumption that a Plan is only valid for a specific Hull) That would remove the possibility of a developer applying a Plan to the wrong Ship.  ie, SelectHull(threat) -> MakePlan(hull) -> SpawnShip(Plan)

Anyway, out-of-scope for this PR; it was just a discussion point.

Reply:
SpawnShip(Plan) would be a lovely future improvement. I have no energy to implement it at the current juncture however :smile: